### DISCUSSION TITLE
Add code coverage functionality

### DISCUSSION THREAD

**Discussion by danieldaeschle** [vlang/v#11742] [Ideas]
http://www.semdesigns.com/Company/Publications/TestCoverage.pdf

**Comment by paulocoghi** [comment]
This is the only feature preventing me from migrating to V as my main language

**Comment by vladimirmyshkovski** [comment]
Is anyone working on this yet? I'm interested in figuring it out and at least starting to do something about it.

**Reply by JalonSolov** [reply]
If you're interested, please do, and submit a PR.

**Comment by prantlf** [comment]
Not the expected implementation, just a result of my search for something, which would work right away :-) An alternative would be writing a V-native code coverage generation.

For example, generating coverage data using gcov for [stringify_test.v](https://github.com/prantlf/v-json/blob/master/stringify_test.v):

```sh
v stringify_test.v -o stringify_test.c
gcc -fprofile-arcs -ftest-coverage -lgc stringify_test.c -o stringify_test
./stringify_test
```

It will write out two binary files:

```txt
stringify_test.gcda
stringify_test.gcno
```

Which need some tool to translate it to humans like [gcovr](https://github.com/gcovr/gcovr), for example:

```sh
mkdir -p coverage
gcovr --html-nested coverage/index.html
```

But with no source maps from C to V, like debugger uses underneath, the pure C analysis includes too much "foreign" code apart from the translated stringify.v:

![obrazek](https://github.com/vlang/v/assets/733193/13fd45ac-8e46-4973-9ed4-b9e1baff996c)

![obrazek](https://github.com/vlang/v/assets/733193/977e4703-1bcd-4d3b-aa5a-7606db776be7)

And a test method:

![obrazek](https://github.com/vlang/v/assets/733193/54c9fa86-e06b-47da-8985-bfe6189d7524)

And the implementation:

![obrazek](https://github.com/vlang/v/assets/733193/4d1385ec-e003-4053-b879-bdfac6449afe)

So, like the debugger can do it, the lines with C functions and structures would need to be mapped to the V source code. The coverage metrics would need to be recomputed according to the V sources using include/exclude configuration based on the V source file and function names. And comments in sources like `/* gcov ignore next 3 */` would need to be recognised. Eventually the recomputed coverage should be in a standard format like `gcov` or `lcov`, so that many existing reporters can be leveraged.
